### ISSUE TITLE
Allow F.pad(mode = 'reflect') when shape == pad

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#52205]
```python
import torch, torch.nn.functional as F

F.pad(torch.rand(2, 1, 256), (256, 0), mode = 'reflect')
#  Argument #4: Padding size should be less than the corresponding input dimension, but got: padding (256, 0) at dimension 2 of input [2, 1, 256]
```
Why can't reflect work by reflecting the whole input?  (and thus including case shape == pad, not only pad < shape)

Also, `zeros` mode gives a very unhelpful error when pad > shape:
```python
F.pad(torch.rand(2, 1, 256), (258, 0), mode = 'zeros')
*** NotImplementedError
```

**Comment by mruberry** [comment]
How does NumPy's reflect padding (even and odd) work?
